multimutex/hash_mutex.go

```diff
@@ -0,0 +1,90 @@
+package multimutex
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/lightningnetwork/lnd/lntypes"
+)
+
+// HashMutex is a struct that keeps track of a set of mutexes with a given hash.
+// It can be used for making sure only one goroutine gets given the mutex per
+// hash.
+type HashMutex struct {
+	// mutexes is a map of hashes to a cntMutex. The cntMutex for
+	// a given hash will hold the mutex to be used by all
+	// callers requesting access for the hash, in addition to
+	// the count of callers.
+	mutexes map[lntypes.Hash]*cntMutex
```

Comment:
can't we just take 4 bytes of the hash and use the existing `uint64` mutex? Alternatively use `interface{}` as the key type.

Reply:
Interesting suggestion. That will work, unless people use non-random hashes (for invoices that are never settled)? There is some cost in terms of understandability. Not sure if it weighs up.

Reply:
Worst case it will just have to wait a bit longer before accessing the DB in those cases. I think it is nice to avoid the code duplication, as long as we don't have generics.

Do avoid the scenario you mention, could also a add a simple method
```go
func (h Hash) Uint64Key() uint64 {
      return uint64(sha256(h[:])[0:4])
```
but maybe that is taking it too far

Reply:
I will leave this thread open for others to comment on.

Reply:
no strong preference fro me, tho i think a hash map would be a useful addition to `multimutex` as a library

Reply:
leaving hash map in then